### PR TITLE
Add `--translate-function-macros` to translate C function-like macros to Rust `const fn`

### PR DESCRIPTION
**Pull request by zRedShift**
## Summary

Partially addresses #753.

Function-like C macros produce no output from bindgen, which means any constant defined through them silently disappears from the generated bindings. This has been a pain point since 2017, affecting Linux ioctls, ALSA, PostgreSQL, cryptsetup, USB device FS macros, and many other headers.

Many of the cases reported in #753 — where users just need the **evaluated constant value** — are now solved by `--clang-macro-fallback` (added separately). That flag evaluates concrete invocations like `#define UI_DEV_CREATE _IO('U', 1)` to `pub const UI_DEV_CREATE: u32 = 21761`. For users who only need the values, that's sufficient.

Some cases remain unsolved by either flag: struct literal macros like raylib's `CLITERAL(Color){ 255, 255, 255, 255 }`, pointer casts like `RTLD_DEFAULT ((void *) 0)`, and other constructs that don't reduce to integer or float expressions.

This PR adds `--translate-function-macros`, which goes further by translating the **macro definitions themselves** into callable Rust `const fn`. The unique value over `--clang-macro-fallback` is programmatic use — calling the macros at compile time with types that don't exist in the original header:

```rust
// Not possible with --clang-macro-fallback alone:
let cmd = _IOR::<MyCustomStruct>(b'U' as u32, 42);
```

The two flags complement each other: `--translate-function-macros` emits the `const fn` definitions, `--clang-macro-fallback` evaluates the concrete invocations to constants.

## Example

```c
#define _IOC(dir,type,nr,size) \
    (((dir) << 30) | ((type) << 8) | (nr) | ((size) << 16))
#define _IO(type,nr)          _IOC(0,(type),(nr),0)
#define _IOR(type,nr,argtype) _IOC(2,(type),(nr),sizeof(argtype))
#define UI_DEV_CREATE         _IO('U', 1)
```

With `--translate-function-macros --clang-macro-fallback`:

```rust
pub const UI_DEV_CREATE: u32 = 21761;           // from --clang-macro-fallback
pub const fn _IOC(dir: u32, ...) -> u32 { ... } // from --translate-function-macros
pub const fn _IOR<argtype>(r#type: u32, nr: u32) -> u32 {
    _IOC(2, (r#type), (nr), core::mem::size_of::<argtype>() as u32)
}
```

## What's supported

Arithmetic, bitwise, shift, and comparison operators; ternary; `sizeof` with generic type params; C-style keyword casts; logical `!`/`&&`/`||` with correct integer truthiness; float literal inference; octal conversion; cross-macro calls with turbofish; forward references (fixpoint resolution); literal suffix stripping (U/UL/ULL/LU/LLU/f/F and MSVC i64/ui64); suffix-aware unsigned type inference matching C promotion rules; target-aware `long` mapping; `func_macro_type` callback; allowlist/blocklist filtering; `#undef`/redefine handling.

Macros that can't be safely translated are skipped; a warning is emitted at `RUST_LOG=warn`.

## Known limitations

1. **Identifiers outside the object-like macro constant map are emitted best-effort** — they get a cast like `(IDENT as vt)`, which works if bindgen generated the identifier in the same output but fails otherwise. Proper fix requires richer symbol resolution from the IR.
2. **Unsigned literal suffix (U) infers u32 or u64** based on value and suffix width, matching C's promotion rules. May choose a wider type than strictly necessary.
3. **Float inference is greedy** — one float literal forces the entire macro to f64. Macros mixing floats with bitwise/modulo/shift are skipped.
4. **sizeof-only macros default to i64** when no constants are referenced. Callers cast the result.

## Test plan

- 24 integration tests covering: basic translation, type inference, allowlists/blocklists, forward references, turbofish, mixed generics, ioctl patterns, ternary with logical ops, float literals, octal, keyword names, typedef casts, arity mismatches, redefines, compiler builtins, and edge cases
- 9 unit tests for suffix stripping, octal conversion, and helpers
- All expectation files verified to compile with `rustc --edition 2021`
- `cargo clippy` and `cargo +nightly fmt --check` clean
- Full `cargo test -p bindgen-tests` passes with same known failures as main

## Disclosure

This PR was developed with significant LLM assistance.

### CODE REVIEW

**Comment by ojeda** [comment]
I just took a quick look at the tests.

Generating a single `ADD(x, y)` for all `i64` can cover a bunch of use cases, especially if this is meant for use in const contexts, but do we need to be more flexible?

The generated `const fn` does `x + y`, just like C. But in general that has different behavior than C, e.g. there is `-Coverflow-checks` in Rust, and in C it is UB or wrapping depending on the sign of the type. For instance, what does the current implementation do if we force `u64`? Does it still output `x + y` or does it switch to wrapping arithmetic?

I guess it all depends if what we are looking for is to mimic C closely or to just expand the operators in a "similar way". This seems more aligned with the latter, if I understand correctly, but in the former case I would naively expect that we would need to have the machinery to take a set of input types and compute the output types like C would do, and to translate each operator etc. (Perhaps you have part of that already). Then we could generate a single function per set of input types (if the user gives the input types manually, either via a flag or `bindgen` generates a macro that then the user has to call to generate the function for the types they want), or perhaps one can use generics/traits to provide a general solution.

So I would perhaps suggest discussing what we want to generate for trivial cases first, and then expand from there, i.e. what kind of use cases we want to cover and how users are going to use it. It would also be nice to have concrete docs on the simple cases that explain how it all works (I see you have an LLM-generated doc, but it doesn't explain the rationale behind the decisions, it doesn't explain how something like `x + y` is supposed to be translated or the different possibilities we have even for such a simple expression, or why we are picking a single type for everything, etc.).

I guess what I am suggesting is to have an RFC to discuss a major feature like this.

By the way, why would `sizeof` not count as a constant using `usize` or similar?

(I don't know the policy on LLM-generated contributors for `bindgen`/Rust, but from quick scrolling it seemed to me some cleanups would be needed).

Thanks for taking a look at this!

File: bindgen/ir/func_macro.rs [hunk 1]
```diff
@@ -0,0 +1,1838 @@
...
+    pub tokens: Vec<OwnedToken>,
+    /// The source file this macro was defined in.
+    pub source_file: Option<String>,
+}
+
+/// Default value type when inference finds nothing.
+pub(crate) const DEFAULT_VALUE_TYPE: &str = "i64";
+
+/// Rust keywords that must be escaped with `r#` when used as identifiers.
+pub(crate) const RUST_KEYWORDS: &[&str] = &[
```
Comment (emilio) [inline bindgen/ir/func_macro.rs:83]: This duplicates already-existing code, please don't.

File: bindgen/codegen/mod.rs [hunk 2]
```diff
@@ -5299,6 +5299,95 @@ pub(crate) fn codegen(
...
+        // Emit translated function-like macros as `const fn` items.
+        // These originate from `#define`s, so they follow var semantics
+        // for filtering: gated on `--generate vars`, filtered by
+        // `--blocklist-var`/`--blocklist-item`, and allowed by
+        // `--allowlist-var`/`--allowlist-item`/`--allowlist-file`.
+        // `--allowlist-type` and `--allowlist-function` do NOT activate
+        // the allowlist gate for function macros — they are orthogonal
+        // (function macros are macro-derived, not C type/function
+        // declarations).
+        if context.options().codegen_config.vars() {
```
Comment (emilio) [inline bindgen/codegen/mod.rs:5311]: This is totally misformatted.

File: bindgen/ir/context.rs [hunk 3]
```diff
@@ -2147,6 +2162,155 @@ If you encounter an error missing from this list, please file an issue or a PR!"
...
+    }
+
+    /// Translate raw function-like macros in a fixpoint pass.
+    ///
+    /// Runs repeated waves until no more progress. Each wave attempts to
+    /// translate all unresolved macros using the current set of already-
+    /// translated macros as context. Forward references between macros
+    /// are resolved naturally: wave 1 translates leaf macros, wave 2
+    /// translates macros that call them, etc.
+    fn translate_function_macros(&mut self) {
```
Comment (emilio) [inline bindgen/ir/context.rs:2191]: Check the config and bail out here instead of checking the caller.

File: bindgen/ir/func_macro.rs [hunk 4]
```diff
@@ -0,0 +1,1838 @@
...
+                _ => {
+                    // Unknown keyword — emit as identifier. The syn
+                    // validation will catch invalid constructs.
+                    let ident = Ident::new(spelling, Span::call_site());
+                    out.extend(quote! { #ident });
+                }
+            },
+
+            CXToken_Comment => {}
+            _ => {}
```
Comment (emilio) [inline bindgen/ir/func_macro.rs:911]: Skipping virtually everything else without error seems extremely footgunny.

**Review by emilio — CHANGES_REQUESTED** [review]
Trying to do CXToken -> Rust expression translation feels extremely sketchy to me. You're re-implementing a lot of the C type checking, expression parsing, C operator precedence, type inference etc in ways that are pretty hard to get right.

There's also very little mention of a lot of the limitations and stuff this just incorrectly skips.

I reckon it's nice that your (Claude's?) code can handle a bunch of the ioctl macros and so on. At the very least some robust error handling would be nice. But in general I wouldn't be confident maintaining this code forward... Can you clarify what is the use case for this vs. the macro fallback stuff that the Linux folks implemented? That seems less whackamole-ish to me.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-bindgen/3350/b7b501feb2642b6ac3796f8c5f2a1461640a2a67..cc086507a3f0a38b76b6848c439e7c86f06a67ae)*
